---
title: "Clustered heatmap of Cryptococcus Wakeup RNAseq timecourse"
author: "Edward Wallace"
date: "2022-11-22"
output: 
  html_document:
    toc: true
---


# Summary

Clustered heatmap of Cryptococcus Wakeup RNAseq timecourse

This analysis does:

- EDIT

It concludes that... EDIT

## Load count data from shared script load_count_data.Rmd

```{r call_load_count_data, child = "load_count_data.Rmd"}
```

```{r load_libraries}
library(ggdendro)
library(vegan)
```


## Set levels to display Condition (Media, Temperature, Time)

```{r set_MTT_levels}
# Order as described in the sample sheet: YPD before RPMI+
# MTT_levels <- 
#  c("YPD_30C_180", 
#    "YPD_25C_0", 
#    "YPD_25C_10", "YPD_25C_30", "YPD_25C_60", "YPD_25C_120",
#    "YPD_37C_10", "YPD_37C_30", "YPD_37C_60", "YPD_37C_120",
#    "RPMIS_25C_10", "RPMIS_25C_30", "RPMIS_25C_60", "RPMIS_25C_120",
#    "RPMIS_37C_10", "RPMIS_37C_30", "RPMIS_37C_60", "RPMIS_37C_120")

# Order with RPMI+ before YPD, moved Exp phase to end 
# MTT_levels <- 
#  c("YPD_25C_0", 
#    "RPMIS_25C_10", "RPMIS_25C_30", "RPMIS_25C_60", "RPMIS_25C_120",
#    "RPMIS_37C_10", "RPMIS_37C_30", "RPMIS_37C_60", "RPMIS_37C_120",
#    "YPD_25C_10", "YPD_25C_30", "YPD_25C_60", "YPD_25C_120",
#    "YPD_37C_10", "YPD_37C_30", "YPD_37C_60", "YPD_37C_120",
#    "YPD_30C_180"
#  )

# Order with RPMI+ before YPD, like timepoints at each temperature grouped 
# MTT_levels <- 
#  c("YPD_25C_0", 
#    "RPMIS_25C_10", "RPMIS_37C_10", 
#    "RPMIS_25C_30", "RPMIS_37C_30",
#    "RPMIS_25C_60", "RPMIS_37C_60",
#    "RPMIS_25C_120", "RPMIS_37C_120",
#    "YPD_25C_10", "YPD_37C_10", 
#    "YPD_25C_30", "YPD_37C_30",
#    "YPD_25C_60", "YPD_37C_60",
#    "YPD_25C_120", "YPD_37C_120",
#    "YPD_30C_180"
#  )

# Order with RPMI+ before YPD, like timepoints at each temperature and media grouped 
MTT_levels <-
  c("YPD_25C_0",
    "RPMIS_25C_10", "RPMIS_37C_10",
    "YPD_25C_10", "YPD_37C_10",
    "RPMIS_25C_30", "RPMIS_37C_30",
    "YPD_25C_30", "YPD_37C_30",
    "RPMIS_25C_60", "RPMIS_37C_60",
    "YPD_25C_60", "YPD_37C_60",
    "RPMIS_25C_120", "RPMIS_37C_120",
    "YPD_25C_120", "YPD_37C_120",
    "YPD_30C_180"
  )
```

# Calculate read count statistics by gene

## Check distribution of counts and baseMean

```{r count_distribution}
count_stats_bygene <- 
  counts_all %>%
  dplyr::select(-Gene, -Length) %>%
  as.matrix() %>%
  tibble(gene = rownames(counts_all),
         count_min = rowMins(.),
         count_mean = rowMeans(.),
         count_max = rowMaxs(.)) %>%
  select( - ".") # removes pipe behaviour leading to inclusion of "."

genelist_mincount_1 <- 
  count_stats_bygene %>%
  filter(count_min >= 1) %>%
  pull(gene)

ggplot(data = count_stats_bygene) +
  geom_density(aes(x=count_min + 1, colour = "min"), 
               size = 1, kernel = "rectangular") +
  geom_density(aes(x=count_max + 1, colour = "max"), 
               size = 1, kernel = "rectangular") +
  geom_density(aes(x=count_mean + 1, colour = "mean"), 
               size = 1, kernel = "rectangular") +
  scale_colour_manual(values = c("min" = "lightblue",
                                 "mean" = "grey20",
                                 "max" = "pink")) + 
  scale_x_log10("log10(count per gene + 1)")
```

# DESeq on a subset of data, only YPD at 1hr

```{r subset, fig.height=8, fig.width=8}
dds_all <- DESeqDataSetFromMatrix(
  countData = counts_all %>%
    dplyr::select(sample_sheet$Code) %>%
    magrittr::set_rownames(counts_all$Gene),
  colData = sample_sheet,
  design = ~ MediumTempTime - 1 ) %>%
  DESeq()

dds_all
resultsNames(dds_all)
```

```{r results}
results(dds_all)
```

## Select MTT-dependent log2FC at each timepoint

```{r tidy_deseq_MTT}
# Create a tidy data frame that contains only the GAT201-dependent log2FC
# in a helpful format for plotting
deseq_df_MTT <- 
  biobroom::tidy.DESeqDataSet(dds_all) %>%
  dplyr::mutate(MediumTempTime = term %>%
           stringr::str_remove("MediumTempTime") %>%
           factor(levels = MTT_levels)) %>%
  dplyr::select(MediumTempTime, gene, baseMean, log2estimate = estimate) %>%
  dplyr::group_by(gene) %>%
  dplyr::mutate(log2FC = log2estimate - mean(log2estimate)) %>%
  dplyr::ungroup()

# restrict to genes with mincount 10
deseq_df_MTT_mincount_1 <- 
  deseq_df_MTT %>%
  filter(gene %in% genelist_mincount_1)

# check we retained all the levels
unique(deseq_df_MTT$MediumTempTime)
```

## Calculate gene expression hierarchical clusters

Here we restrict to genes with at least 11 counts in each sample.
This list was calculated earlier as `genelist_mincount_11`. 

This avoids a problem with numeric errors in calculations for all genes, that probably came from genes with low/zero counts.

```{r hclust_genes}
log2FC_MTT_m1_wide <-
  deseq_df_MTT_mincount_1  %>%
  select(gene, MediumTempTime, log2FC) %>%
  pivot_wider(id_cols = gene, names_from = MediumTempTime, values_from = log2FC) %>%
  select(c("gene", MTT_levels))

preferred_gene_ordering <- 
  log2FC_MTT_m1_wide %>% 
  dplyr::pull(YPD_25C_0)

# Function to allow calculation of distance by correlation
# not exactly as euclidean distance
# decided not to use but keeping for this commit
distcort <- function(x) {
  as.dist(1 - cor(t(x)))
}
  
hclust_log2FC <- 
  log2FC_MTT_m1_wide %>%
  dplyr::select(-gene) %>%
  as.matrix() %>%
  magrittr::set_rownames(log2FC_MTT_m1_wide$gene) %>%
  # head(n= 100) %>%
  # dist() %>%
  distcort() %>%
  hclust(method = "average") %>%
  reorder(wts = preferred_gene_ordering)
```

## Plot clustering dendrogram by itself

```{r plot_dendrogram_log2FC_m1}
# dendrogram data
hclust_log2FC_dendrodf <- 
  hclust_log2FC %>% dendro_data()


plot_dendrogram_log2FC_m1 <- 
  ggplot(data = hclust_log2FC_dendrodf$segments) + 
  geom_segment(aes(x = x, xend = xend, y = y, yend = yend)) +
  # geom_label(data = hclust_log2FC_dendrodf$labels %>%
  #              filter(label %in% c("CNAG_06963", "CNAG_03143")),
  #            aes(x = x, y = y, label = label)) +
  theme_void() +
  scale_y_reverse() +
  coord_flip(expand = c(0,0), clip = "off")
plot_dendrogram_log2FC_m1
```

## Plot the log2 fold-change in same order as clustering dendrogram

```{r plot_log2FC_m1_dorder, fig.height = 7, fig.width = 5}

deseq_df_MTT_mincount_1_dorder <- 
  deseq_df_MTT_mincount_1 %>%
  mutate(gene_dorder = 
           factor(gene,
                  levels = log2FC_MTT_m1_wide$gene[hclust_log2FC$order],
                  ordered = TRUE)
         )

plot_log2FC_m1_dorder <- 
  ggplot(data = deseq_df_MTT_mincount_1_dorder,
         aes(x = MediumTempTime, fill = log2FC, y = gene_dorder)) +
  geom_tile() +
  scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow",
                       limits = c(-3, 3), oob = scales::squish,
                       guide = guide_colorbar(nbin = 36, raster = FALSE)) +
  coord_cartesian(expand = c(0,0), clip = "off") + 
  theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
        panel.grid.major = element_blank(),
        axis.line = element_blank(),
        axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
  labs(x = "Temperature, Media, Time",
       y = "Gene, ordered by clustering")

plot_log2FC_m1_dorder

# ggsave(filename = "../results/clustered_heatmap_CryptoWakeupMediumTemperatureTime_tall.pdf", 
#       plot_log2FC_m1_dorder + theme(axis.text.y = element_text(size = 2)),
#       width = 6, height = 72, limitsize = FALSE)
```

```{r heatmap_genes_select}
genes_translation <- c(RPS3 = "CNAG_00116", 
                       RPL2 = "CNAG_05232", 
                       TEF1 = "CNAG_06125")
genes_stationary <- c(CQS1 = "CNAG_03012", 
                      HSP12 = "CNAG_03143", 
                      BLP2 = "CNAG_06347")
genes_heat <- c(SSA1 = "CNAG_06443", 
                HSP90 = "CNAG_06150", 
                # HSP10 = "CNAG_03892"
                TSA3 = "CNAG_06917")
genes_GAT201 <- c(GAT201 = "CNAG_01551", 
                  GAT204 = "CNAG_06762", 
                  LIV3 = "CNAG_05835")

genes_select <- c(genes_translation, genes_stationary, 
                  genes_heat, genes_GAT201)

ggplot(data = deseq_df_MTT_mincount_1_dorder %>% 
         dplyr::filter(gene %in% genes_select),
       aes(x = MediumTempTime, fill = log2FC, y = gene_dorder)) +
  geom_tile() +
  scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow",
                       limits = c(-3, 3), oob = scales::squish) +
  scale_y_discrete(breaks = genes_select, labels = names(genes_select)) +
  coord_cartesian(expand = c(0,0), clip = "off") + 
  theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
        panel.grid.major = element_blank())

# plot_log2FC_m1_dorder + 
#   geom_text(deseq_df_MTT_mincount_1_dorder %>% 
#          dplyr::filter(gene %in% genes_select,
#                        MediumTempTime == "YPD_25C_0"),
#          mapping = aes(x=0, y = gene_dorder, label = str_remove(gene, "CNAG_0")),
#          size = 2, colour = "red")
```


## Make a nicer x-axis for sample labels

This section programs up x-axis labels split up by GAT201, Time, Serum

It's ugly code to get a pretty result.

```{r nice_MTT_labels, fig.height = 1, fig.width = 5}
MTT_df <- 
  tibble(MTT = as_factor(MTT_levels),
                 x = as.numeric(MTT)) %>%
  tidyr::separate(MTT, c("Medium","Temperature","Time"), remove = FALSE) %>% 
  dplyr::mutate(Phase = Time %>%
           forcats::fct_collapse(Stat = "0",
                                 Lag  = c("10","30","60","120"),
                                 Exp = "180"),
         Medium = str_replace(Medium, "RPMIS", "RPMI+"),
         Temperature = str_replace(Temperature, "C", "°C"),
         MediumTemp = paste(Medium, Temperature, sep = " "),
         Time = as.integer(Time)
         )

MTT_MTime_df <- MTT_df %>%
  dplyr::filter(MTT != "YPD_25C_0") %>%
  dplyr::group_by(Medium,Time) %>%
  dplyr::summarise(xmean = mean(x), xmin = min(x), xmax = max(x),
                   .groups = "keep")

MTT_MTime_df <- MTT_df %>%
  dplyr::filter(MTT != "YPD_25C_0") %>%
  dplyr::group_by(Medium,Time) %>%
  dplyr::summarise(xmean = mean(x), xmin = min(x), xmax = max(x),
                   .groups = "keep")

MTT_Time_df <- MTT_df %>%
  # dplyr::filter(MTT != "YPD_25C_0") %>%
  dplyr::group_by(Time) %>%
  dplyr::summarise(xmean = mean(x), xmin = min(x), xmax = max(x),
                   .groups = "keep") 

plot_nice_MTT_labels_texttemp <- 
  ggplot(data = MTT_df  %>%
           dplyr::filter(MTT != "YPD_25C_0") ) +
  geom_text(aes(x = x, label = Temperature), y = 3, size = 2.6) + 
  geom_text(data = MTT_MTime_df, aes(x = xmean, label = Medium), y = 2, size = 2.9) + 
  geom_segment(data = MTT_MTime_df,
               aes(x = xmin, xend = xmax), y = 2.5, yend = 2.5) +
  geom_text(data = MTT_Time_df, aes(x = xmean, label = Time), y = 1) + 
  geom_segment(data = MTT_Time_df,
               aes(x = xmin, xend = xmax), y = 1.5, yend = 1.5) +
  scale_y_continuous(limits = c(0.5,3.5),
                     breaks = c(1,2,3),
                     labels = c("Time (mins)", "Medium", "Temperature")) + 
  scale_x_continuous(limits = c(0.5, max(MTT_df$x + 0.5))) + 
  coord_cartesian(expand = c(0,0), clip = "off") + 
  theme_nothing() + 
  theme(axis.text.y = element_text())

plot_nice_MTT_labels <- 
  ggplot(data = MTT_df  %>%
           dplyr::filter(MTT != "YPD_25C_0") ) +
  geom_point(aes(x = x, colour = MediumTemp), y = 3, size = 2.5) + 
  scale_colour_manual(
    "Condition",
    # purpleorange
    values = c("RPMI+ 37°C" = "#5e3c99",
               "RPMI+ 25°C" = "#b2abd2",
               "YPD 37°C" = "#e66101",
               "YPD 25°C" = "#fdb863",
               "YPD 30°C" = "darkgreen")) +
  geom_text(data = MTT_MTime_df, aes(x = xmean, label = Medium), y = 2, size = 2.9) + 
  geom_segment(data = MTT_MTime_df,
               aes(x = xmin, xend = xmax), y = 2.5, yend = 2.5) +
  geom_text(data = MTT_Time_df, aes(x = xmean, label = Time), y = 1) + 
  geom_segment(data = MTT_Time_df,
               aes(x = xmin, xend = xmax), y = 1.5, yend = 1.5) +
  scale_y_continuous(limits = c(0.5,3.5),
                     breaks = c(1,2,3),
                     labels = c("Time (mins)", "Medium", "Temperature")) + 
  scale_x_continuous(limits = c(0.5, max(MTT_df$x + 0.5))) + 
  coord_cartesian(expand = c(0,0), clip = "off") + 
  theme_nothing() + 
  theme(axis.text.y = element_text())

plot_nice_MTT_labels

# the next bit plots the y-axis labels as their own ggplot, right-aligned
# so that we can get everything to align nicely in the plot_grid later
plot_nice_MTT_yaxis <- 
  ggplot(data = tibble(y = c(1,2,3),
                     label = c("Time (mins)", "Medium", "Temperature")),
         aes(y=y, label = label)) +
  geom_text(hjust = 1, x = 1) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0.5, 3.5),
                  expand = c(0,0), clip = "off")  + 
  theme_nothing() 
```


# Combined plot of dendrogram and heatmap

This combines the previous plot panels into a single plot, showing the dendrogram aligned with the clustered heatmap, and with nice x-axis labels.

```{r combine_dendrogram_heatmap, fig.height = 6, fig.width = 7}
plot_dh_nolegend <- 
  plot_grid(plot_dendrogram_log2FC_m1 + 
              theme(plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_log2FC_m1_dorder +
            theme(axis.title.y = element_blank(),
                  axis.title.x = element_blank(),
                  axis.text.x = element_blank(),
                  axis.ticks.x = element_blank(),
                  legend.position = "none",
                  plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_nice_MTT_yaxis  + 
              theme(plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_nice_MTT_labels + 
            theme(axis.text.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm")),
          ncol = 2,
          align = "hv",
          rel_widths = c(1,5),
          rel_heights = c(7,1)
          )

plot_grid(get_legend(plot_log2FC_m1_dorder + 
                       labs(fill = "mRNA log2 fold-change") +
                       theme(legend.box.margin = margin(0, 0, 0, 2, unit = "cm"), 
                     legend.position = "top")),
          plot_dh_nolegend,
          ncol = 1,
          rel_heights = c(1,7))

# ggsave(filename = "../results/clustered_heatmap_GAT201ConditionTime.png", width = 6, height = 6)

```

### Extract and plot clusters

```{r extract_cluster_functions}

make_clustered_df <- function(k = 2, 
                              names_df = log2FC_MTT_m1_wide,
                              hclust_results = hclust_log2FC) {
      tibble(gene = names_df$gene,
         gene_dorder = 
           factor(gene,
                  levels = names_df$gene[hclust_results$order],
                  ordered = TRUE),
         y = as.integer(gene_dorder),
         cluster = hclust_results %>% 
           cutree(k = k) %>% 
           factor())
}

get_cluster_from_clustered_df <- function(gene_bait, 
                                          clustered_df,
                                          output = c("list", "df")) {
  # find cluster containing the gene_bait (gene of interest)
  cluster_number <- clustered_df %>%
    dplyr::filter(gene == gene_bait) %>%
    dplyr::pull(cluster)
  # return all of that cluster
  clustered_df_filtered <- 
    clustered_df %>%
    dplyr::filter(cluster == cluster_number)
  if (output == "df") {
    return(clustered_df_filtered)
  } else if (output == "list") {
    return(clustered_df_filtered$gene)
  } 
}

make_cluster_with_gene <- function(gene_bait,
                                   k = 2, 
                                   names_df = log2FC_MTT_m1_wide,
                                   hclust_results = hclust_log2FC,
                                   output = c("list", "df")) {
  clustered_df <- 
    make_clustered_df(k = k, 
                      names_df = names_df,
                      hclust_results = hclust_results)
  get_cluster_from_clustered_df(gene_bait, clustered_df)
}

plot_clustered_genelist <- 
  function(genelist, data = deseq_df_MTT_mincount_1_dorder) {
    ggplot(data = data %>% 
             dplyr::filter(gene %in% genelist),
           aes(x = MediumTempTime, fill = log2FC, y = gene_dorder)) +
      geom_tile() +
      scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow",
                           limits = c(-3, 3), oob = scales::squish) +
      scale_y_discrete(breaks = genelist, labels = names(genelist)) +
      coord_cartesian(expand = c(0,0), clip = "off") + 
      theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
            panel.grid.major = element_blank())
    
  }
```

```{r extract_clusters}
genecluster_CQS1 <- make_cluster_with_gene(gene = "CNAG_03012", k = 300)

write_lines(x = genecluster_CQS1, 
            file = here::here("results",
                              "geneclusters",
                              "genecluster_stationary_k300.txt"))

plot_genecluster_CQS1 <- plot_clustered_genelist(genecluster_CQS1)


genecluster_SSA1 <- make_cluster_with_gene(gene = "CNAG_06443", k = 300)

write_lines(x = genecluster_SSA1, 
            file = here::here("results",
                              "geneclusters",
                              "genecluster_heat_k300.txt"))

plot_genecluster_SSA1 <- plot_clustered_genelist(genecluster_SSA1)


genecluster_TEF1 <- make_cluster_with_gene(gene = "CNAG_06125", k = 300)

write_lines(x = genecluster_TEF1, 
            file = here::here("results",
                              "geneclusters",
                              "genecluster_translation_k300.txt"))

plot_genecluster_TEF1 <- plot_clustered_genelist(genecluster_TEF1)


genecluster_TSA3 <- make_cluster_with_gene(gene = "CNAG_06917", k = 500)

write_lines(x = genecluster_TSA3, 
            file = here::here("results",
                              "geneclusters",
                              "genecluster_TSA3_k500.txt"))

plot_genecluster_TSA3 <- plot_clustered_genelist(genecluster_TSA3)


genecluster_GAT201 <- make_cluster_with_gene(gene = "CNAG_01551", k = 700)

write_lines(x = genecluster_GAT201, 
            file = here::here("results",
                              "geneclusters",
                              "genecluster_GAT201_k700.txt"))

plot_genecluster_GAT201 <- plot_clustered_genelist(genecluster_GAT201)
```

```{r collect_clusters}
hclust_colour_df_grouped <-
  tibble(gene = log2FC_MTT_m1_wide$gene,
         gene_dorder =
           factor(gene,
                  levels = log2FC_MTT_m1_wide$gene[hclust_log2FC$order],
                  ordered = TRUE),
         y = as.integer(gene_dorder),
         cluster_select = fct_collapse(gene,
                                       CQS1   = genecluster_CQS1,
                                       HSP    = genecluster_SSA1,
                                       GAT201 = genecluster_GAT201,
                                       TSA3   = genecluster_TSA3,
                                       RPs    = genecluster_TEF1,
                                       other_level = "other") %>%
           forcats::fct_recode(NULL = "other") 
         ) %>%
  group_by(cluster_select) %>%
  summarise(n = n(), ymin = min(y), ymax = max(y), ymean = mean(y))
```

```{r plot_clusters}
plot_cluster_select <- 
  ggplot(data = hclust_colour_df_grouped %>%
           dplyr::filter(!is.na(cluster_select))) +
  geom_rect(aes(ymin = ymin, ymax = ymax
                #, fill = cluster_select
                ),
            fill = "grey50",
            xmin = -0.1, xmax = 0) +
  # geom_text(aes(y = ymean, 
  #               colour = cluster_select, 
  #               label = cluster_select),
  #           x = 0.1, hjust = 0) +
  coord_cartesian(xlim = c(-0.1,0), 
                  ylim = c(min(hclust_colour_df_grouped$ymin),
                           max(hclust_colour_df_grouped$ymax)),
                  expand = c(0,0), clip = "off") +
  #scale_colour_brewer() + scale_fill_brewer() +
  theme_nothing()
  
plot_cluster_select
```

```{r combine_dendrogram_heatmap_clusterlabels, fig.height = 6, fig.width = 7}
plot_dh_cluster_nolegend <- 
  plot_grid(plot_dendrogram_log2FC_m1 + 
              theme(plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_log2FC_m1_dorder +
            theme(axis.title.y = element_blank(),
                  axis.title.x = element_blank(),
                  axis.text.x = element_blank(),
                  axis.ticks.x = element_blank(),
                  legend.position = "none",
                  plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_cluster_select + 
              theme(plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_nice_MTT_yaxis + 
              theme(plot.margin = unit(c(0, 0, 0, 0), "cm")),
          plot_nice_MTT_labels + theme(axis.text.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm")),
          NULL,
          ncol = 3,
          align = "hv",
          rel_widths = c(1,5,0.1),
          rel_heights = c(7,1)
          )

plot_grid(get_legend(plot_log2FC_m1_dorder + 
                       labs(fill = "mRNA log2 fold-change") +
                       theme(legend.box.margin = margin(0, 0, 0, 2, unit = "cm"), 
                             legend.key.width = unit(0.25,"in"),
                             legend.position = "top")),
          plot_dh_cluster_nolegend,
          ncol = 1,
          rel_heights = c(1,7))

ggsave(filename = "../results/clustered_heatmap_GAT201ConditionTime.pdf", width = 5, height = 5)

```


To make those clusters containing genes of interest

- cuttree to make the clusters
- search the cluster that gene of interest is in
- select all the genes from that cluster
- plot them
- vary number or height of clusters until it makes sense

## Plot fold-change of example genes as line plot

```{r plot_log2FC_line, fig.height = 3, fig.width = 4}
plot_log2FC_line <- function(mygenes, 
                             datadf = deseq_df_MTT,
                             labdf = MTT_df,
                             ylimits = c(-2.8,2.8)) {
  mydata <- 
    deseq_df_MTT %>%
    dplyr::filter(gene %in% mygenes, 
                  MediumTempTime != "YPD_30C_180") %>%
    dplyr::select(MTT = MediumTempTime, gene, log2FC) %>%
    dplyr::left_join(labdf, by = "MTT")
  ggplot(data = mydata, 
         aes(x = Time, y = log2FC,
             colour = MediumTemp, group = MediumTemp)) +
    geom_hline(yintercept = 0, size = 0.5, colour = "grey90") +
    geom_line(size = 1) +
    scale_y_continuous("mRNA log2\nfold-change",
                       limits = ylimits, breaks = seq(-4,4),
                       expand = c(0.01,0.01), oob = scales::squish) +
    scale_x_continuous("Time (mins.)", 
                       breaks = c(0,10,30,60,120),
                       expand = c(0.01,0.01)) + 
    scale_colour_manual(
      "Condition",
      # purpleorange
      values = c("RPMI+ 37°C" = "#5e3c99",
                 "RPMI+ 25°C" = "#b2abd2",
                 "YPD 37°C" = "#e66101",
                 "YPD 25°C" = "#fdb863")) +
    theme(plot.title = element_text(size = rel(1)),
          plot.margin = unit(c(0, 0.5, 0, 0.1), "cm"))
}

lineplot_CQS1 <- 
  plot_log2FC_line(mygenes = "CNAG_03012") + labs(title = "CQS1")
lineplot_CQS1

lineplot_SSA1 <- 
  plot_log2FC_line(mygenes = "CNAG_06443") + labs(title = "SSA1")

lineplot_GAT201 <- 
  plot_log2FC_line(mygenes = "CNAG_01551") + labs(title = "GAT201")

lineplot_TSA3 <- 
  plot_log2FC_line(mygenes = "CNAG_06917",
                   ylimits = c(-3,5)) + labs(title = "TSA3")

lineplot_TEF1 <- 
  plot_log2FC_line(mygenes = "CNAG_06125") + labs(title = "TEF1")
```

## Facet plot of example genes for main figure

```{r plot_log2FC_line_faceted, fig.height = 6, fig.width = 3}

plot_mygenes_facet <- function(mygenes,
                               mygenenames, 
                               deseq_df = deseq_df_MTT,
                               meta_df = MTT_df) {
  mygenes_df <- 
    tibble(gene  = mygenes %>% as_factor(), 
           name = mygenenames %>% as_factor()) %>%
    dplyr::left_join(deseq_df) %>%
    dplyr::filter(MediumTempTime != "YPD_30C_180") %>%
    dplyr::select(MTT = MediumTempTime, gene, name, log2FC) %>%
    dplyr::left_join(meta_df, by = "MTT")
  
  ggplot(data = mygenes_df, 
         aes(x = Time, y = log2FC,
             colour = MediumTemp, group = MediumTemp)) +
    geom_hline(yintercept = 0, size = 0.5, colour = "grey90") +
    geom_line(size = 1) +
    scale_y_continuous("mRNA log2 fold-change",
                       # limits = ylimits,
                       breaks = seq(-4,4),
                       expand = c(0.01,0.2)) +
    scale_x_continuous("Time (mins.)", 
                       breaks = c(0,10,30,60,120),
                       expand = c(0.01,0.01)) + 
    scale_colour_manual(
      "Condition",
      # purpleorange
      values = c("RPMI+ 37°C" = "#5e3c99",
                 "RPMI+ 25°C" = "#b2abd2",
                 "YPD 37°C" = "#e66101",
                 "YPD 25°C" = "#fdb863")) + 
    guides(colour = guide_legend(
      title.position = "top",
      ncol = 2)) +
    facet_grid(name~., scales = "free", space = "free") +
    panel_border(color = "black", size = 0.5) +
    theme(strip.text.y = element_text(angle = 0),
          legend.position = "none",
          plot.margin = unit(c(0.1, 0, 0.1, 0), "in"))
}

plot_mygenes_facet(
  mygenes = c("CNAG_03012", "CNAG_06443", "CNAG_01551", "CNAG_06917", "CNAG_06125"),
  mygenenames = c("CQS1", "SSA1", "GAT201", "TSA3", "TEF1")
)

ggsave(filename = "../results/log2FC_line_multipanel_clusterselect.pdf",
       width = 2.5, height = 5)
```


# TO DO LIST

- Write figure legend
- Select clusters containing genes of interest - DONE
  - Plot some mini-heatmaps - DONE
  - Output genelists for GO analysis - DONE
  - Do GO analysis either manually or via FungiDB API?
- Put the figure together
  - Improve the mini-heatmap formatting (remove stuff)
  - Move the log2FC legend
  - Plot the mini-heatmaps next to the big heatmap
  - Adjust sizes until it looks nice
  - Display GO analysis on figure, somehow