---
title: "Quality control and PCA of CryptoWakeup RNAseq"
author: "Edward Wallace"
date: "2022-08-31"
output: 
  html_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)
library(GGally)
library(here)
library(DESeq2)
library(biobroom)

theme_set(theme_cowplot(font_size = 12) +
    theme(# panel.grid.major = element_line(colour = "grey80", size = 0.5),
          strip.background = element_blank())
    )

# Count header columns from featurecounts
count_header_columns <- c("Geneid", "Chr", "Start",
                          "End", "Strand", "Length")

# unwanted string in filename; may need to be edited
unwanted_filename_string_bam <- "Lall_aln.bam"

```

# Summary

Quality control of CryptoWakeup RNA-seq data after read processing.

This analysis does:

- Heatmap of counts showing log-scale correlations.
- PCA plot showing how samples are similar to one another.

It concludes that :

- biological replicates are reproducible.
- principal components capture meaningful factors: growth time, medium and temperature.
- the variance structure is complex, probably due to the complex multifactorial design.


# Load and organise raw count data

## Load the raw counts

This is the output of featurecounts, one integer per gene.

```{r load_counts}
counts_raw <- 
  here::here("results", "counts", "counts.txt") %>%
  readr::read_tsv(comment = "#",
                  col_types = "ccccciiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii")
counts_raw
```

## Load the sample sheet

The sample sheet describes the sample information, filenames, as well as sequence and tag names relevant to library prep.

```{r load_samplesheet}
sample_sheet <- 
  here::here("input", "experiment",
             "CryptoWakeUpSampleSheetPlusTagsAndRawFilenames.txt") %>%
  readr::read_tsv(comment = "#") %>%
  dplyr::mutate(BamFileName = paste(P7, 
                                    TagID, 
                                    unwanted_filename_string_bam, 
                                    sep = "_")
  )

sample_sheet 
```

Next make the abbreviated sample sheet containing only the experimental design information, which is summarized in the "Code" variable that uniquely identifies each sample.

```{r samplesheet_abbrev}
sample_sheet_abbreviated <- 
  sample_sheet %>%
  dplyr::select(SampleID, Code, Medium, Temp, Time, Rep)

Codelevels = sample_sheet$Code
```

## Organise the counts using the sample sheet

```{r organise_counts}
counts_all <- 
  counts_raw %>%
  dplyr::select(sample_sheet$BamFileName)  %>%
  magrittr::set_names(sample_sheet$Code) %>%
  magrittr::set_rownames(counts_raw$Geneid)

counts_all
```

# Check reproducibility by paired scatter plot in log10-plus-1-space

_Note: This is set to `eval=FALSE` as runs very slowly for large datasets._ 
_Tried running it and it was very hard to interpret._

```{r paired_scatter, fig.height=16, fig.width=16, eval = FALSE}
counts_all %>%
  magrittr::add(1) %>%
  log10() %>%
  GGally::ggpairs()
```


# Check reproducibility by correlation heatmap in log10-plus-1 space

```{r paired_heatmap, fig.height=8, fig.width=10}
cordf <- 
  counts_all %>%
  magrittr::add(1) %>%
  log10() %>%
  cor() %>%
  as.data.frame() %>%
  mutate(Sample1 = names(.)) %>%
  pivot_longer(cols = -Sample1, 
               names_to = "Sample2", 
               values_to = "R") %>%
  mutate(Sample1 = factor(Sample1, levels = Codelevels),
         Sample2 = factor(Sample2, levels = rev(Codelevels)))

ggplot(data = cordf, aes(x = Sample1, y = Sample2, fill = R)) +
  geom_tile() +
  coord_equal() +
  scale_fill_gradient(low = "white", high = "darkblue") +
  theme(axis.text.x = element_text(angle = 90))
```

This is a complex experiment with 36 samples so the plot is hard to interpret.
The biological replicates (first 18 samples rep A, next 18 rep B) are well-correlated.
The visible 4x4 squares on the diagonal indicate that later timepoints in the same media and temperature are quite well correlated with each other, better than with timepoint 1 (10min).


# Normalize counts by regularized log (rlog) with DESeq2

The [rlog function from DESeq2](https://rdrr.io/bioc/DESeq2/man/rlog.html) "transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size."

## Construct DESeq dataset object

```{r dds_counts_all}
dds_all <- DESeqDataSetFromMatrix(countData = counts_all,
                                  colData = sample_sheet_abbreviated,
                                  design = ~ Code)
```

# Do the rlog transformation

```{r rlog_dds_all}
rlog_all <- rlog(dds_all)
head(assay(rlog_all))
```

# Check reproducibility by principal component analysis (PCA) of rlog

## Calculate PCs from the rlog output

```{r calculate_pca_rlog}
# calculate principal components of rlog, after extracting from the dataset
pca_rlog <- rlog_all %>%
  assay() %>%
  t() %>%
  prcomp()

# convert principal components to data frame
pcdf_rlog <- bind_cols(
  as_tibble(colData(rlog_all)),
  as_tibble(pca_rlog$x)
)

pcdf_rlog
```

## Calculate proportion of variance explained by each PC


```{r calculate_propvar}
propvar_rlog_df <- tibble(
  PC = seq.int(1L, ncol(pca_rlog$x) ),
  prop_var = pca_rlog$sdev^2/sum(pca_rlog$sdev^2)
)
```

## Plot PCA percentage variance explained to find how many PCs to pay attention to.

```{r plot_percentvar, fig.width = 3, fig.height = 2}
plot_percentvar_rlog <- 
  ggplot(data = propvar_rlog_df, 
         aes(x = PC, y = prop_var)) +
  geom_col(fill = "blue") +
  scale_x_continuous("principal component",
                     limits = c(0.4,10.6), 
                     # hack limits to include the bars and a little space.
                     breaks = 1L:10L,
                     expand = c(0,0)) + 
  scale_y_continuous("prop. of variance", expand = c(0,0))
plot_percentvar_rlog
```

Here about 32% var in PC1, 27% in PC2, then several more components between 10% and 2% of variance.

## Plot PC1 vs PC2 with default values

EDIT: the colour and other aesthetics may need to be further edited to highlight features of the data

```{r plot_PC12_default, fig.height=4, fig.width=7}
ggplot(data = pcdf_rlog,
       aes(colour = interaction(Medium,Temp), shape = factor(Time))
       ) +
  geom_point(aes(x = PC1, y = PC2)) 
```

Overall this argues that the biological replicates are clustering together in PC space.


## Plot PC1 vs PC2, 3, 4, also with default values

```{r plot_PC1234_text, fig.height=8, fig.width=8}
# This could be wraped in a function
pc_baseplot <- 
  ggplot(data = pcdf_rlog,
         aes(colour = interaction(Medium,Temp), 
             label = Code)
  ) + 
  geom_hline(yintercept = 0, size = 0.5, colour = "grey80") +
  geom_vline(xintercept = 0, size = 0.5, colour = "grey80") +
  theme(legend.position = "none")

pcplot_12 <- pc_baseplot + geom_text(aes(x = PC1, y = PC2))
pcplot_13 <- pc_baseplot + geom_text(aes(x = PC1, y = PC3))
pcplot_14 <- pc_baseplot + geom_text(aes(x = PC1, y = PC4))
pcplot_23 <- pc_baseplot + geom_text(aes(x = PC2, y = PC3))
pcplot_24 <- pc_baseplot + geom_text(aes(x = PC2, y = PC4))
pcplot_34 <- pc_baseplot + geom_text(aes(x = PC3, y = PC4))

pcplot_legend <- get_legend(pcplot_12 + theme(legend.position = "right"))

plot_grid(pcplot_12, NULL, pcplot_legend, 
          pcplot_13, pcplot_23, NULL, 
          pcplot_14, pcplot_24, pcplot_34)
```


- PC1 vs PC2 Stationary phase (YC0A/B) cluster distinctly, and the 10min samples (with "1" in them) are all similar to each other but different from later timepoints.
- PC2 separates RPMI from YPD samples at later timepoints, looking at the left side of the PC1/PC2 plot.
- PC3 mostly separates temperature, with 25C (cold) at the top at 37C (hot) at the bottom.
- PC4 separates stationary phase (YC0A/B) from everything else.
- A better choice of colours and display might make this clearer.

# Write out counts and TPMs (length-normalized counts) by sample code

## Write out counts with sample codes as column names

```{r write_counts_codes}
bind_cols(
  counts_raw %>% 
    dplyr::select(Gene = Geneid, Length = Length),
  counts_raw %>%
    dplyr::select(sample_sheet$BamFileName)  %>%
    magrittr::set_names(sample_sheet$Code) 
) %>%
  readr::write_tsv(
    file = here::here("results", "summaries", "counts_bysamplecode.txt"))
```

## Calculate gene-by-gene estimates in transcripts per million

```{r calculate_tpms}
calculate_tpms <- function(counts, effLengths) {
  norm_counts <- counts / effLengths
  norm_counts / sum(norm_counts) * 1e6
}

# fragment length = 100 is a rough approximation from PE75 reads.
fragment_length <- 100

tpms_values <- 
  counts_raw %>%
  dplyr::select(sample_sheet$BamFileName)  %>%
  magrittr::set_names(sample_sheet$Code) %>%
  dplyr::mutate(
    across(.fns = calculate_tpms, 
           effLengths = pmin(counts_raw$Length - fragment_length, 1))
  )

tpms_all <- bind_cols( counts_raw %>% 
                         dplyr::select(Gene = Geneid, Length),
                       tpms_values)
tpms_all
```

```{r write_tpms_codes}
tpms_all %>%
  dplyr::mutate(
    across(where(is.double), round, digits = 3)
  ) %>%
  readr::write_tsv(
    file = here::here("results", "summaries", "tpms_bysamplecode.txt"))
```

